src/process.cr

```diff
@@ -85,19 +128,351 @@ class Process
+      if error
+        @wait_count += 1
+        spawn { copy_io(process_error, error, channel, close_src: true) }
+      else
+        @error = process_error
+      end
+    end
+
+    @pid = Process.fork_internal(run_hooks: false) do
+      begin
+        Process.exec_internal(
+          command,
+          argv,
+          env,
+          clear_env,
+          exec_io_for(fork_input || input),
+          exec_io_for(fork_output || output),
+          exec_io_for(fork_error || error),
+          chdir
+        )
+      rescue ex
+        ex.inspect_with_backtrace STDERR
+        LibC._exit 127 # TODO: Compiler bug: ensure isn't run when an exception is raised and catched here
+      ensure
+        LibC._exit 127
+      end
+    end
+
+    @waitpid_future = Event::SignalChildHandler.instance.waitpid(pid)
+
+    fork_input.try &.close
+    fork_output.try &.close
+    fork_error.try &.close
+  end
+
+  protected def initialize(@pid)
+    @waitpid_future = Event::SignalChildHandler.instance.waitpid(pid)
+    @wait_count = 0
   end
+
+  # See `Process.kill`
+  def kill(sig = Signal::TERM)
+    Process.kill sig, @pid
+  end
+
+  # Waits for this process to complete and closes any pipes.
+  def wait : Process::Status
+    close_io @input # only closed when a pipe was created but not managed by copy_io
+
+    @wait_count.times do
+      ex = channel.receive
+      raise ex if ex
+    end
+    @wait_count = 0
+
+    @waitpid_future.get
+  ensure
+    close
+  end
+
+  # Whether the process is still registered in the system.
+  # Note that this returns true for processes in the zombie or similar state.
+  def exists?
+    !terminated?
+  end
+
+  # Whether this process is already terminated.
+  def terminated?
+    @waitpid_future.completed? || !Process.exists?(@pid)
+  end
+
+  # Closes any pipes to the child process.
+  def close
+    close_io @input
+    close_io @output
+    close_io @error
+  end
+
+  # :nodoc:
+  protected def self.prepare_argv(command, args, shell)
+    if shell
+      command = %(#{command} "${@}") unless command.includes?(' ')
+      shell_args = ["-c", command, "--"]
+
+      if args
+        unless command.includes?(%("${@}"))
+          raise ArgumentError.new(%(can't specify arguments in both, command and args without including "${@}" into your command))
+        end
+
+        {% if flag?(:freebsd) %}
+          shell_args << ""
+        {% end %}
+
+        shell_args.concat(args)
+      end
+
+      command = "/bin/sh"
+      args = shell_args
+    end
+
+    argv = [command.to_unsafe]
+    args.try &.each do |arg|
+      argv << arg.to_unsafe
+    end
+    argv << Pointer(UInt8).null
+
+    {command, argv}
+  end
+
+  private def channel
+    @channel ||= Channel(Exception?).new
+  end
+
+  private def needs_pipe?(io)
+    io.nil? || (io.is_a?(IO) && !io.is_a?(IO::FileDescriptor))
+  end
+
+  private def exec_io_for(io)
+    # TODO: Compiler bug: The union doesn't properly degrade to IO::FileDescriptior|Bool
```

Comment:
Well, it brings something back, I can no longer reproduce it complaining about `nil` instead of `MemoryIO`:

```diff
diff --git a/src/process.cr b/src/process.cr
index 5d9bff5..c683032 100644
--- a/src/process.cr
+++ b/src/process.cr
@@ -235,9 +235,9 @@ class Process
           argv,
           env,
           clear_env,
-          fork_input || input,
-          fork_output || output,
-          fork_error || error,
+          (fork_input || input).not_nil!,
+          (fork_output || output).not_nil!,
+          (fork_error || error).not_nil!,
           chdir
         )
       rescue ex
@@ -360,7 +360,7 @@ class Process
   end
 
   # :nodoc:
-  protected def self.exec_internal(command : String, argv, env, clear_env, input, output, error, chdir)
+  protected def self.exec_internal(command : String, argv, env, clear_env, input : Bool|IO::FileDescriptor, output : Bool|IO::FileDescriptor, error  : Bool|IO::FileDescriptor, chdir)
     reopen_io(input, STDIN, "r")
     reopen_io(output, STDOUT, "w")
     reopen_io(error, STDERR, "w")
```
`make spec` errors with:

```
in ./src/process.cr:233: no overload matches 'Process.exec_internal' with types String, Array(Pointer(UInt8)), Nil, Bool, (IO::FileDescriptor | MemoryIO), IO::FileDescriptor, IO::FileDescriptor, Nil
Overloads are:
 - Process.exec_internal(command : String, argv, env, clear_env, input : Bool | IO::FileDescriptor, output : Bool | IO::FileDescriptor, error : Bool | IO::FileDescriptor, chdir)
Couldn't find overloads for these types:
 - Process.exec_internal(command : String, argv : Array(Pointer(UInt8)), env : Nil, clear_env : Bool, input : MemoryIO, output : IO::FileDescriptor, error : IO::FileDescriptor, chdir : Nil)

        Process.exec_internal(
                ^~~~~~~~~~~~~
```

Since casting to a sub-union isn't possible, this was my solution to tell Crystal what the actually possible types are at this point.

```diff
diff --git a/src/process.cr b/src/process.cr
index 5d9bff5..e56defa 100644
--- a/src/process.cr
+++ b/src/process.cr
@@ -235,9 +235,9 @@ class Process
           argv,
           env,
           clear_env,
-          fork_input || input,
-          fork_output || output,
-          fork_error || error,
+          (fork_input || input).as(Bool|IO::FileDescriptor),
+          (fork_output || output).as(Bool|IO::FileDescriptor),
+          (fork_error || error).as(Bool|IO::FileDescriptor),
           chdir
         )
       rescue ex
@@ -360,7 +360,7 @@ class Process
   end
 
   # :nodoc:
-  protected def self.exec_internal(command : String, argv, env, clear_env, input, output, error, chdir)
+  protected def self.exec_internal(command : String, argv, env, clear_env, input : Bool|IO::FileDescriptor, output : Bool|IO::FileDescriptor, error  : Bool|IO::FileDescriptor, chdir)
     reopen_io(input, STDIN, "r")
     reopen_io(output, STDOUT, "w")
     reopen_io(error, STDERR, "w")
```
actually crashes the compiler now that I try it.